File: mobileapp/src/pairing/deviceName.test.ts
import {deviceLabel, displayDeviceName} from './deviceName';

// The pair roster's job is letting you tell YOUR devices apart well enough to revoke the
// right one. It was failing twice over: every entry read `gtmux • iPhone` — a "gtmux"
// prefix inside gtmux's own roster (nothing in that list is not a gtmux device) over a
// word true of every iPhone ever made. Two paired phones were indistinguishable.

describe('a device names itself with what it actually knows', () => {
  test('the idiom, and no product prefix', () => {
    expect(deviceLabel('ios', 'phone')).toBe('iPhone');
    expect(deviceLabel('ios', 'pad')).toBe('iPad');
    expect(deviceLabel('ios', 'phone')).not.toMatch(/gtmux/i);
  });

  // The OS version reaches the Mac on every request (X-Gtmux-Client) and the roster row
  // prints it underneath the name. A name carrying it said the same thing twice, and its
  // copy froze at pairing while the line below stayed current.
  test('never the OS version: the row already shows the live one', () => {
    for (const v of [deviceLabel('ios', 'phone'), deviceLabel('ios', 'pad'), deviceLabel('android')]) {
      expect(v).not.toMatch(/iOS|Android\s*[0-9]|[0-9]/);
    }
  });

  test('a missing part is left out, never rendered as "undefined"', () => {
    expect(deviceLabel('ios', undefined)).toBe('iPhone');
    expect(deviceLabel('android')).toBe('Android');
    for (const v of [deviceLabel('ios'), deviceLabel('android'), deviceLabel('')]) {
      expect(v).not.toMatch(/undefined|null|NaN/);
      expect(v).not.toBe('');
    }
  });
});

describe('roster rows tidy up without a re-pair', () => {
  test('the legacy prefix is stripped in every form it was written', () => {
    expect(displayDeviceName('gtmux • iPhone')).toBe('iPhone');
    expect(displayDeviceName('gtmux · iPad')).toBe('iPad');
    expect(displayDeviceName('gtmux iPhone')).toBe('iPhone');
    expect(displayDeviceName('GTMUX • iPhone · iOS 18.5')).toBe('iPhone');
  });

  // A roster paired under the older rule reads "iPad · iOS 26.6.1" over a second line
  // that already says "iOS 26.6.1 · …" (seen 2026-09-20). The frozen copy comes off here,
  // so the list tidies itself without a re-pair.
  test('an OS version the name carried is dropped: the row prints the live one', () => {
    expect(displayDeviceName('iPad · iOS 26.6.1')).toBe('iPad');
    expect(displayDeviceName('iPhone · iOS 18.5')).toBe('iPhone');
    expect(displayDeviceName('Android 34')).toBe('Android 34'); // not a "· version" tail
  });

  test('a name of their own is untouched', () => {
    expect(displayDeviceName('dev-mbp.local')).toBe('dev-mbp.local');
    expect(displayDeviceName('ccy')).toBe('ccy');
    expect(displayDeviceName('Lin · iPad')).toBe('Lin · iPad');
  });

  test('a bare generic kind is title-cased, not left as a lowercase word', () => {
    expect(displayDeviceName('browser')).toBe('Browser');
    expect(displayDeviceName('terminal')).toBe('Terminal');
  });

  test('a row never renders empty', () => {
    // A device legitimately named after the tool keeps something to show.
    expect(displayDeviceName('gtmux')).toBe('gtmux');
    expect(displayDeviceName('')).toBe('—');
  });
});

File: internal/app/devicename_test.go
package app

import "testing"

// The pair roster's job is letting you tell YOUR devices apart well enough to revoke the
// right one, and every entry read "gtmux • iPhone" — a product prefix inside gtmux's own
// roster (device-roster-naming).
func TestDeviceDisplayNameDropsTheLegacyPrefix(t *testing.T) {
	for raw, want := range map[string]string{
		"gtmux • iPhone":            "iPhone",
		"gtmux · iPad":              "iPad",
		"gtmux iPhone":              "iPhone",
		"GTMUX • iPhone · iOS 18.5": "iPhone",
		// a bare generic kind is title-cased so it doesn't read as a lowercase word
		"browser":  "Browser",
		"terminal": "Terminal",
		// untouched
		"Lin · iPad":    "Lin · iPad",
		"dev-mbp.local": "dev-mbp.local",
		"ccy":           "ccy",
		// a device legitimately named after the tool keeps something to show
		"gtmux": "gtmux",
		"":      "",
	} {
		if got := deviceDisplayName(raw); got != want {
			t.Errorf("deviceDisplayName(%q) = %q; want %q", raw, got, want)
		}
	}
}

// The roster prints the version a device reports on every request beside its name, and
// keeps it current. A name carrying its own copy said it twice and froze it at pairing:
// one roster read "iPad · iOS 26.6.1" over "iOS 26.6.1 · 127.0.0.1 · last seen 19h ago"
// (2026-09-20).
func TestDeviceDisplayNameDropsAnEchoedOSVersion(t *testing.T) {
	cases := map[string]string{
		"iPad · iOS 26.6.1":         "iPad",
		"iPhone · iOS 18.5":         "iPhone",
		"gtmux • iPhone · iOS 18.5": "iPhone",
		"Lin · iPad":                "Lin · iPad", // a name of their own keeps its parts
		"Android 34":                "Android 34", // not a "· version" tail
		"iOS 26.6.1":                "iOS 26.6.1", // nothing else left: keep it
	}
	for in, want := range cases {
		if got := deviceDisplayName(in); got != want {
			t.Errorf("deviceDisplayName(%q) = %q, want %q", in, got, want)
		}
	}
}

// A revoke that can only name an id tells the person who typed it nothing, and after the
// roster names stopped carrying a frozen OS version ("iPad · iOS 26.6.1"), two iPads look
// alike. The line names the device with what separates it from its twin.
func TestRevokedWhatNamesTheDeviceNotJustItsID(t *testing.T) {
	devs := []deviceListEntry{
		{ID: "e816142782cee09e", Name: "iPad · iOS 26.6.1", Platform: "iOS 26.6.1", LastIP: "127.0.0.1"},
		{ID: "c3bb6e4c5a4253e5", Name: "gtmux • iPhone", Platform: "iOS 26.6.2"},
		{ID: "32a54fd0878c6e80", Name: "browser"},
	}
	for _, c := range []struct{ id, want string }{
		{"e816142782cee09e", "iPad (iOS 26.6.1 · 127.0.0.1) · e816142782cee09e"},
		{"c3bb6e4c5a4253e5", "iPhone (iOS 26.6.2) · c3bb6e4c5a4253e5"},
		{"32a54fd0878c6e80", "Browser · 32a54fd0878c6e80"},
		{"nosuch", "nosuch"}, // nothing to add: the id alone still tells the truth
	} {
		if got := labelForRevoke(devs, c.id); got != c.want {
			t.Errorf("labelForRevoke(%q) = %q, want %q", c.id, got, c.want)
		}
	}
}
